### PR TITLE
Implement #172 Add OIDC access token validation support

### PR DESCRIPTION
**Pull request by hengsin**
This is ready for testing but might goes through further changes down the road due to project requirement is still fluid at this point.

### CODE REVIEW

**Comment by hengsin** [comment]
anyone free to review this ?

File: com.trekglobal.idempiere.rest.api/keycloak/Testing Steps.txt [hunk 1]
```diff
@@ -0,0 +1,76 @@
+1. Install keycloak at port 9080 with password admin/admin
+- sudo docker run -p 9080:8080 -e KEYCLOAK_ADMIN=admin -e KEYCLOAK_ADMIN_PASSWORD=admin quay.io/keycloak/keycloak:21.1.1 start-dev
```
Comment (CarlosRuiz-globalqss) [inline com.trekglobal.idempiere.rest.api/keycloak/Testing Steps.txt:2]: suggesting to change this to:

```
docker run -p 9080:8080 -e KEYCLOAK_ADMIN=admin -e KEYCLOAK_ADMIN_PASSWORD=admin quay.io/keycloak/keycloak:latest start-dev
```

**Comment by CarlosRuiz-globalqss** [comment]
Hi @hengsin - I tested this successfully with the test steps provided, but I have two doubts:

1 - the refresh mechanism didn't work, I set the expiration time in keycloak to 1 minute, and after that minute passed postman answered with an "401 Unauthorized" error.  Is that going to be implemented later?

2 - I didn't test, but I would like to check this connected with an API gateway (the recommended approach), but we can do that and fix later.

I'm OK with pushing this as the tests went fine, but I think the auto-refresh is a must for any application, so would like to know if this is planning to solve that shortcome from this plugin in future.

**Comment by hengsin** [comment]
No, auto refresh is the responsibility of the client. The rest-api service is just the receiver of the access token and the only responsibility is to validate the token received.

File: com.trekglobal.idempiere.rest.api/keycloak/Testing Steps.txt (hunk 1 above)
Comment (hengsin) [reply com.trekglobal.idempiere.rest.api/keycloak/Testing Steps.txt:2]: Exact version help to ensure the testing instruction is valid (some details could change between version). Anyway, the instruction is just for testing and not something for production use.

**Comment by CarlosRuiz-globalqss** [comment]
> No, by design, token refresh is the responsibility of the client/caller, i.e the party that initiate the authentication and obtain the access and refresh token. The rest-api service is just the receiver of the access token and the only responsibility is to validate the token received.

Right, but, how to client/caller would refresh the token?  We don't have a refresh token API, or must the caller get the refresh token calling directly an API in keycloak?

**Comment by hengsin** [comment]
> > No, by design, token refresh is the responsibility of the client/caller, i.e the party that initiate the authentication and obtain the access and refresh token. The rest-api service is just the receiver of the access token and the only responsibility is to validate the token received.
> 
> Right, but, how to client/caller would refresh the token? We don't have a refresh token API, or must the caller get the refresh token calling directly an API in keycloak?

The access token is issue by Keycloak, not iDempiere. The client/caller must call keycloak api  to refresh it.

**Comment by CarlosRuiz-globalqss** [comment]
Hi @hengsin - I pushed this according to your answers.

Although having this in REST and not in iDempiere configures an undesired mismatch, the user would need to login in REST with one password and another in iDempiere.

One more question, please:

> The access token is issue by Keycloak, not iDempiere. The client/caller must call keycloak api to refresh it.

In order to test this, can you please explain how that would work with keycloak?

**Comment by hengsin** [comment]
> Hi @hengsin - I pushed this according to your answers.
> 
> Although having this in REST and not in iDempiere configures an undesired mismatch, the user would need to login in REST with one password and another in iDempiere.
> 
> One more question, please:
> 
> > The access token is issue by Keycloak, not iDempiere. The client/caller must call keycloak api to refresh it.
> 
> In order to test this, can you please explain how that would work with keycloak?

This is build for integration or third party client use case where user account is created for rest api call only and not for login to iDempiere Web Client. Of course, it is possible to use the same SSO for iDempiere web client after we have merge the SSO PR from Deepak.

When you call protocol/openid-connect/token, the response comes back with jsonData.access_token and jsonData.refresh_token.
 
As per the thread below, the refresh_token can then be used to get a new access token later:
https://stackoverflow.com/questions/51386337/refresh-access-token-via-refresh-token-in-keycloak

**Comment by CarlosRuiz-globalqss** [comment]
Thanks @hengsin - I just did the test and it worked fine:

I added the example for refresh to the postman file with
https://github.com/bxservice/idempiere-rest/commit/717f5567c097206d5adb1844f8eee76d90727553

**Pull request merged**